be/src/exprs/agg/approx_top_k.h

```diff
@@ -0,0 +1,471 @@
+                right = mid;
+            }
+        }
+        if (counters[left].count == min_count) {
+            return left;
+        }
+        DCHECK_GT(left, 0);
+        return left - 1;
+    }
+    // Maintain the ordering of the counters, counters with smaller count are put in the front
+    // This is used to speed up the lookup of the min counter
+    void _maintain_ordering(const size_t idx) {
+        DCHECK_LT(idx, unused_idx);
+        size_t start = idx, end = idx;
+        if (start > 0 && counters[start].count < counters[start - 1].count) {
+            while (start > 0 && counters[start].count < counters[start - 1].count) {
+                std::swap(counters[start].value, counters[start - 1].value);
+                std::swap(counters[start].count, counters[start - 1].count);
+                start--;
+            }
+        } else if (end < unused_idx - 1 && counters[end].count > counters[end + 1].count) {
+            while (end < unused_idx - 1 && counters[end].count > counters[end + 1].count) {
+                std::swap(counters[end].value, counters[end + 1].value);
+                std::swap(counters[end].count, counters[end + 1].count);
+                end++;
+            }
+        }
+        if (start < end) {
+            for (size_t i = start; i <= end; i++) {
+                table.erase(counters[i].value);
+            }
+            for (size_t i = start; i <= end; i++) {
+                table[counters[i].value] = &counters[i];
+            }
+        }
+        DCHECK(std::is_sorted(counters.begin(), counters.begin() + unused_idx, cmp));
+        [[maybe_unused]] auto check = [this]() -> bool {
+            // Check index
+            for (size_t i = 0; i < counter_num; i++) {
+                if (i != counters[i]._index) {
+                    return false;
+                }
+            }
+            // Check table
+            static EQ eq;
+            for (auto& [k, v] : table) {
+                if (!eq(k, v->value)) {
+                    return false;
+                }
+            }
+            return true;
+        };
+        DCHECK(check());
+    }
+
+    template <bool IsDeepCopy>
+    CppType _copy(MemPool* mem_pool, const CppType& value) {
+        if constexpr (IsDeepCopy && IsSlice<CppType>) {
+            uint8_t* pos = mem_pool->allocate(value.size);
+            std::memcpy(pos, value.data, value.size);
+            return Slice{pos, value.size};
+        } else {
+            return value;
+        }
+    }
+};
+
+template <LogicalType LT, typename T = RunTimeCppType<LT>>
+class ApproxTopKAggregateFunction final
+        : public AggregateFunctionBatchHelper<ApproxTopKState<LT>, ApproxTopKAggregateFunction<LT, T>> {
+public:
+    using CppType = RunTimeCppType<LT>;
+    using InputColumnType = RunTimeColumnType<LT>;
+
+    static constexpr int32_t MAX_COUNTER_NUM = 100000;
+
+    std::pair<int32_t, int32_t> get_k_and_counter_num(FunctionContext* ctx) const {
+        int32_t k = 5;
```

Comment:
why using 5 as initial value?

Reply:
databricks and snowflake both use 5 as default.

Reply:
> databricks and snowflake both use 5 as default.

got it…but better to define a static const variable for it

Reply:
Yeah, updated.